src/Access/AccessRights.cpp

```diff
@@ -408,6 +408,64 @@ struct AccessRights::Node
 
     friend bool operator!=(const Node & left, const Node & right) { return !(left == right); }
 
+    bool contains(const Node & other)
+    {
+        if (!flags.contains(other.flags))
+            return false;
+
+        /// Let's assume that the current node has the following rights:
+        ///
+        /// SELECT ON *.* TO user1;
+        /// REVOKE SELECT ON system.* FROM user1;
+        /// REVOKE SELECT ON mydb.* FROM user1;
+        ///
+        /// And the other node has the rights:
+        ///
+        /// SELECT ON *.* TO user2;
+        /// REVOKE SELECT ON system.* FROM user2;
+        ///
+        /// First, we check that each child from the other node is present in the current node:
+        ///
+        /// SELECT ON *.* TO user1;  -- checked
+        /// REVOKE SELECT ON system.* FROM user1; -- checked
+        std::unordered_set<std::string_view> seen;
+        if (other.children)
+        {
+            for (const auto & [name, node] : *other.children)
+            {
+                seen.emplace(name);
+                const auto & child = tryGetChild(name);
+                if (child == nullptr)
+                {
+                    if (!flags.contains(node.flags))
+                        return false;
+                }
+                else
+                {
+                    if (!child->contains(node))
+                        return false;
+                }
+            }
+        }
+
+        if (!children)
+            return true;
+
+        /// Then we check that each of our children has no other rights revoked.
+        ///
+        /// REVOKE SELECT ON mydb.* FROM user1; -- check failed, returning false
+        for (const auto & [name, node] : *children)
+        {
+            if (seen.contains(name))
```

Comment:
We can use the following condition instead of this one:
```
if (other.children && other.children->contains(name))
    continue; /// we've already checked matched children of `*this` and `other`
```
so we don't really need this extra set `seen`.

Reply:
I can do this, but if it was a real tree, using `unordered_set` would be much faster (luckily, our tree depth is no greater than 3, so it doesn't really matter in this case.

Reply:
`children` is declared as `std::unique_ptr<std::unordered_map<std::string_view, Node>>`, and
`seen` is declare as std::unordered_set<std::string_view>`. Why should searching in `seen` be much faster?

Reply:
Oh, OK, nevermind